### ISSUE TITLE
Maximum call stack size exceeded

### ISSUE THREAD

**Issue by evil-shrike** [jrburke/r.js#869]
After moving from 2.1.8 to 2.1.20 my rjs-based build started failing with RangeError: Maximum call stack size exceeded error.
Here's stack from console:
```
Error: RangeError: Maximum call stack size exceeded
    at Object.lang.deeplikeCopy (D:\Work\R-n-D\XFW3_WebClient-tools\grunt-croc-requirejs\node_modules\requirejs\bin\r.js:2819:32)

    at D:\Work\R-n-D\XFW3_WebClient-tools\grunt-croc-requirejs\node_modules\requirejs\bin\r.js:28575:19
    at D:\Work\R-n-D\XFW3_WebClient-tools\grunt-croc-requirejs\node_modules\requirejs\bin\r.js:3052:39
    at D:\Work\R-n-D\XFW3_WebClient-tools\grunt-croc-requirejs\node_modules\requirejs\bin\r.js:2992:25
    at Function.prim.nextTick (D:\Work\R-n-D\XFW3_WebClient-tools\grunt-croc-requirejs\node_modules\requirejs\bin\r.js:28405:9)
    at Object.p.errback (D:\Work\R-n-D\XFW3_WebClient-tools\grunt-croc-requirejs\node_modules\requirejs\bin\r.js:2991:26)
    at Object.p.callback (D:\Work\R-n-D\XFW3_WebClient-tools\grunt-croc-requirejs\node_modules\requirejs\bin\r.js:2977:23)
    at Object.p.promise.then (D:\Work\R-n-D\XFW3_WebClient-tools\grunt-croc-requirejs\node_modules\requirejs\bin\r.js:3031:23)
    at build (D:\Work\R-n-D\XFW3_WebClient-tools\grunt-croc-requirejs\node_modules\requirejs\bin\r.js:28532:12)
    at runBuild (D:\Work\R-n-D\XFW3_WebClient-tools\grunt-croc-requirejs\node_modules\requirejs\bin\r.js:30476:17)
    at Object.context.execCb (D:\Work\R-n-D\XFW3_WebClient-tools\grunt-croc-requirejs\node_modules\requirejs\bin\r.js:1929:33)
```
But the stack is unclear as it's async code. So actual execution path looks like:
`optimize` ->  `flattenModule` -> `createOverrideConfig(config, overide)` -> `deeplikeCopy(config)`

Starting from `deeplikeCopy(config)` infinite recursion starts as iteration reaches "defined" object.
`deeplikeCopy` tries to clone config object with the following path:
`config` -> `config.modules` -> `config.modules` -> `config.modules[0]` -> `config.modules[0].layer` -> `config.modules[0].layer.context` -> `config.modules[0].layer.context.defined` -> `config.modules[0].layer.context.defined.handlebars` -> `config.modules[0].layer.context.defined.handlebars.default`

`config.modules[0].layer.context.defined.handlebars.default` - Is a cycle (it refers `HandlebarsEnvironment` instance)

`defined` field contains a map with modules names. All values are undefined. But except several modules. They are rjs plugins I'm using to load modules. It's standard `text` and some my plugins (to load/compile Handlerbars template and combine css). Also the map contains module `handlebars` with a value - `HandlerbarsEnvironment` object. Cloning this object causes infinite recursion.

I understand that having objects in `defined` map isn't OK and probably it's caused by my customization of build process.  
But it's hard for me to understand how to avoid infinite recursion. The `config` object which I'm passing to rjs's `optimize` method doesn't contains any objects with cycles. 
Its `modules` field looks like:
```
"modules":[
{"name":"vendor-layer","exclude":[]},
{"name":"lib-layer","exclude":["vendor-layer"]},
{"name":"main","exclude":["vendor-layer","lib-layer"],"override":{}},
{"name":"report-main","exclude":["vendor-layer","lib-layer"],"override":{}}
]
```
Here's my plugins:
* https://github.com/CrocInc/grunt-croc-requirejs/blob/master/rjs/xhtmpl.js
* https://github.com/CrocInc/grunt-croc-requirejs/blob/master/rjs/xcss.js

**Comment by jrburke** [comment]
The deep copy was copying too much, unrelated to what was needed. I fixed this in master if you want to confirm the fix before 2.1.21 is complete (may be another day or two):

https://raw.githubusercontent.com/jrburke/r.js/master/dist/r.js

**Comment by evil-shrike** [comment]
The fix works fine for me. Thanks!
